lib/prelude.dx

```diff
@@ -895,79 +895,59 @@ instance {a n} [Ord a] Ord (n=>a)
 '## Subset class
 
 interface Subset a_sub a
-  inject' : a_sub -> a
-  project' : a -> Maybe a_sub
+  inject'              : a_sub -> a
+  can_project' a_sub a : a -> Bool
+  unsafe_project'      : a -> a_sub
+
+def can_project {a} (a_sub:Type) [Subset a_sub a] (x:a) : Bool =
+  can_project' a_sub a x
 
 def project {a} (a_sub:Type) [Subset a_sub a] (x:a) : Maybe a_sub =
-  project' x
+  case can_project a_sub x of
+    True  -> Just $ unsafe_project' x
+    False -> Nothing
+
+def unsafe_project {a} (a_sub:Type) [Subset a_sub a] (x:a) : a_sub =
+  unsafe_project' x
 
 def inject {a_sub} (a:Type) [Subset a_sub a] (x:a_sub) : a =
   inject' x
 
 instance {a b c} [Subset a b, Subset b c] Subset a c
   inject' = \x. inject c $ inject b x
-  project' = \x. case project b x of
-    Nothing -> Nothing
-    Just y -> project a y
-
-instance {a b} Subset a (a|b)
-  inject' = \x. Left x
-  project' = \x. case x of
-    Left  x -> Just x
-    Right x -> Nothing
-
-instance {a b} Subset a (b|a)
-  inject' = \x. Right x
-  project' = \x. case x of
-    Left  x -> Nothing
-    Right x -> Just x
+  can_project' = \x. case can_project b x of
+    False -> False
+    True -> can_project a $ unsafe_project b x
+  unsafe_project' = \x. unsafe_project a $ unsafe_project b x
 
 instance {q:Type} {i:q} [Ix q] Subset (RangeFrom q i) q
   inject' = \(UnsafeMkRangeFrom j).
     unsafe_from_ordinal _ $ j + ordinal i
-  project' = \j.
-    j' = ordinal j
-    i' = ordinal i
-    if j' < i'
-      then Nothing
-      else Just $ UnsafeMkRangeFrom $ unsafe_nat_diff j' i'
+  can_project' = \j. ordinal j >= ordinal i
+  unsafe_project' = \j.
+    UnsafeMkRangeFrom $ unsafe_nat_diff (ordinal j) (ordinal i)
 
 instance {q:Type} {i:q} [Ix q] Subset (RangeFromExc q i) q
   inject' = \(UnsafeMkRangeFromExc j).
     unsafe_from_ordinal _ $ j + ordinal i + 1
-  project' = \j.
-    j' = ordinal j
-    i' = ordinal i
-    if j' <= i'
-      then Nothing
-      else Just $ UnsafeMkRangeFromExc $ unsafe_nat_diff j' (i' + 1)
+  can_project' = \j. ordinal j > ordinal i
+  unsafe_project' = \j.
+    UnsafeMkRangeFromExc $ unsafe_nat_diff (ordinal j) (ordinal i + 1)
 
 instance {q:Type} {i:q} [Ix q] Subset (RangeTo q i) q
   inject' = \(UnsafeMkRangeTo j). unsafe_from_ordinal _ j
-  project' = \j.
-    j' = ordinal j
-    i' = ordinal i
-    if j' > i'
-      then Nothing
-      else Just $ UnsafeMkRangeTo j'
+  can_project' = \j. ordinal j <= ordinal j
```

Comment:
Should there be an `ordinal i` here somewhere?  `ordinal j <= ordinal j` with two `j`s doesn't look right.  And a unit test that would catch the difference between `i` and `j`?

Reply:
Yes, this is clearly a bug, and shows that I didn't put in enough tests.

Reply:
I restored all the old `project` instances back to their old selves.